### PR TITLE
Cimpl6 import

### PR DESCRIPTION
**Pull request by AmartC**
Model changes needed to support CIMPL6.  The two most significant changes to the model are:

* Replaced `code`, `Coding`, and `CodeableConcept` with new `concept` primitive
* Removed `RefValue` in favor of using a field's "entry-ness" to determine if it should be a reference

All unit tests should pass.  For information on running w/ CIMPL6 source see:
https://github.com/standardhealth/shr-cli/pull/207

### CODE REVIEW

File: lib/dataElementListener.js [hunk 1]
```diff
@@ -76,12 +85,18 @@ class DataElementImporter extends SHRDataElementParserListener {
...
+    if (docHeader.version()) {
+      const version = docHeader.version();
+      const major = parseInt(version.WHOLE_NUMBER()[0], 10);
+      const minor = parseInt(version.WHOLE_NUMBER()[1], 10);
+      this._currentGrammarVersion = new Version(major, minor);
+
+      logger.debug({shrId: ns, version: this._currentGrammarVersion.toString()}, 'Start importing data element namespace');
+    }
+    else {
+      logger.error('Grammar declaration not found. ERROR_CODE 11???');
```
Comment (cmoesel) [inline lib/dataElementListener.js]: Needs a real error code.

File: lib/dataElementListener.js [hunk 2]
```diff
@@ -153,23 +222,178 @@ class DataElementImporter extends SHRDataElementParserListener {
...
+          if (this._currentDef.value instanceof ChoiceValue) {
+            let marked = false;
+            let options = this._currentDef.value.options;
+            options.forEach(function(element) {
+              if (ctx.elementWithConstraint().getText().includes(`[${element.identifier.name}]`)) {
+                marked = true;
+              }
+            });
+            if(!marked) {
+              logger.error('Choice value constrained without specifying the specific choice. ERROR_CODE:11???');
```
Comment (cmoesel) [inline lib/dataElementListener.js]: Needs a real error code

File: lib/dataElementListener.js [hunk 3]
```diff
@@ -153,23 +222,178 @@ class DataElementImporter extends SHRDataElementParserListener {
...
+      }
+      else {
+        const match = this._currentDef.fields.find(f => f.effectiveIdentifier && f.effectiveIdentifier.equals(field.identifier));
+        if (match) {
+          field.constraints.forEach(c => match.addConstraint(c));
+        }
+        else {
+          const secondMatch = this._currentDef.fields.find(f => f.identifier && f.identifier.equals(field.identifier));
+          if (secondMatch) {
+            logger.error('Constraint refers to previous identifier. ERROR_CODE:11???');
```
Comment (cmoesel) [inline lib/dataElementListener.js]: Needs a real error code

File: lib/dataElementListener.js [hunk 4]
```diff
@@ -234,100 +458,30 @@ class DataElementImporter extends SHRDataElementParserListener {
...
+      const value = new ChoiceValue();
+      for (const t of typeCtxArr) {
+        value.addOption(this.processType(t, 1, 1));
       }
+      value.setMinMax(1,1);
       return value;
     }
+    let match = typeCtxArr[0].getText().match(/\d+\.\.(\d+|\*)/);
+    if (match) {
+      logger.error('Value should not be declaring cardinality. ERROR:11???');
```
Comment (cmoesel) [inline lib/dataElementListener.js]: Needs a real error number

File: lib/dataElementListener.js [hunk 5]
```diff
@@ -234,100 +458,30 @@ class DataElementImporter extends SHRDataElementParserListener {
...
+    }
+    else if (typeCtxArr[0].getText().length == 0) {
+      logger.error('Missing a value element. ERROR:11???');
```
Comment (cmoesel) [inline lib/dataElementListener.js]: Needs a real error number

File: package.json [hunk 6]
```diff
@@ -17,18 +17,21 @@
     "lint:fix": "./node_modules/.bin/eslint . --fix"
   },
   "dependencies": {
-    "antlr4": "~4.6.0"
+    "antlr4": "~4.6.0",
+    "dtrace-provider": "^0.8.7"
```
Comment (cmoesel) [inline package.json]: I don't think this should be here.  `dtrace-provider` is a transitive dependency, not a direct dependency -- and putting it here doesn't seem to make the `dtrace-provider` stack traces go away anyway.

File: yarn-error.log [hunk 7]
```diff
@@ -0,0 +1,1123 @@
+Arguments:
```
Comment (cmoesel) [inline yarn-error.log]: Please don't check in `yarn-error.log`.  Remove.

File: lib/dataElementListener.js [hunk 8]
```diff
@@ -62,9 +64,16 @@ class DataElementImporter extends SHRDataElementParserListener {
...
+
   enterDoc(ctx) {
     // Process the namespace
-    const ns = ctx.docHeader().namespace().getText();
+    let ns = '';
+    if (ctx.docHeader().namespace()) {
+      ns = ctx.docHeader().namespace().getText();
+    }
+    else {
+      logger.error('Namespace declaration not found. ERROR_CODE:11???');
```
Comment (cmoesel) [inline lib/dataElementListener.js]: This needs a real error code. Just try to find the next number that's not used.  This page may help, although I'm not sure if it is up to date: https://github.com/standardhealth/shr-cli/wiki/Error-Message-Documentation#all-error-codes

File: lib/dataElementListener.js [hunk 9]
```diff
@@ -153,23 +222,178 @@ class DataElementImporter extends SHRDataElementParserListener {
...
   enterField(ctx) {
-    const field = this.processCountAndTypes(ctx.count(), ctx.fieldType());
-    if (field instanceof IncompleteValue && field.identifier && field.identifier.isValueKeyWord) {
-      if (this._currentDef.value) {
-        logger.error('Elements cannot use "Value:" modifier and specify "_Value" field at same time. ERROR_CODE:11024');
-        return;
+    if (ctx.propertyField()) {
+      const field = this.processCountAndTypes(ctx.propertyField().count(), [ctx.propertyField().propertyFieldType()]);
+      if (this._currentDef.fields.some(f => f.identifier && f.identifier.equals(field.identifier))) {
+        logger.error('Property "%s" already exists. ERROR_CODE:11???', field.identifier.name);
```
Comment (cmoesel) [inline lib/dataElementListener.js]: This needs a real error code

File: lib/dataElementListener.js [hunk 10]
```diff
@@ -30,6 +30,8 @@ class DataElementImporter extends SHRDataElementParserListener {
     this._currentFile = '';
     // The currently active definition (DataElement)
     this._currentDef = null;
+    // The currently active CodeSystem definition (if applicable)
+    this._codeSystemDefs = [];
```
[the remark this answers is not in the document]
Comment (AmartC) [reply lib/dataElementListener.js]: yeah looks like an error.

[the remark this answers is not in the document]
Comment (AmartC) [reply lib/dataElementListener.js]: I'll remove it.

**Comment by cmoesel** [comment]
Approved BUT DO NOT MERGE until we are ready to merge everything.  (I will merge when ready).

**Comment by cmoesel** [comment]
Given that this has 218 commits, with a bunch of merge commits, I'm going to squash-and-merge in order to keep some sanity in the commit history.

**Pull request merged**